redbot/core/commands/dpy_reimplements.py

```diff
@@ -0,0 +1,237 @@
+from __future__ import annotations
+
+import inspect
+import functools
+from typing import TypeVar, Callable, Awaitable, Coroutine, Union, Type, TYPE_CHECKING, List
+
+import discord
+from discord.ext import commands as dpy_commands
+
+from ..i18n import Translator
+from .context import Context
+from .commands import Command
+
+
+_ = Translator("nah", __file__)
+
+
+"""
+Anything here is either a reimplementation or re-export
+of a discord.py funtion or class with more lies for mypy
+"""
+
+__all__ = [
+    "check",
+    # "check_any",  # discord.py 1.3
+    "guild_only",
+    "dm_only",
+    "is_nsfw",
+    "has_role",
+    "has_any_role",
+    "bot_has_role",
+    "bot_has_any_role",
+    "when_mentioned_or",
+    "cooldown",
+    "when_mentioned",
+]
+
+_T_Call = TypeVar("_T_Call", bound=Command)  # small lie for mypy
+
+SameObjectDecorator = Callable[[_T_Call], _T_Call]
+Predicate = Callable[[Context], Awaitable[bool]]
+SyncPredicate = Callable[[Context], bool]
+CheckType = Type[Callable]
+
+
+def check(predicate: Union[SyncPredicate, Predicate]) -> SameObjectDecorator:
+    def decorator(func: _T_Call) -> _T_Call:
+
+        if isinstance(func, Command):
+            func.checks.append(predicate)
+        else:
+            if not hasattr(func, "__commands_checks__"):
+                func.__commands_checks__ = []
+
+            func.__commands_checks__.append(predicate)
+
+        return func
+
+    if inspect.iscoroutinefunction(predicate):
+        setattr(decorator, "predicate", predicate)
+    else:
+
+        @functools.wraps(predicate)
+        async def wrapper(ctx):
+            return predicate(ctx)
+
+        setattr(decorator, "predicate", wrapper)
```

Comment:
Why is this not using dot notation?

Reply:
Intentional. Mypy doesn't appreciate dynamic access to attributes which aren't declared